storage/src/main/java/org/apache/kafka/storage/internals/epoch/LeaderEpochFileCache.java

```diff
@@ -525,10 +525,13 @@ private void writeToFile() {
         }
     }
 
-    private void writeToFileForTruncation() {
+    private void writeIfDirExists() {
         lock.readLock().lock();
         try {
-            checkpoint.writeForTruncation(epochs.values());
+            // Write epoch entries under the read lock to avoid older epoch entries overwriting the newer epoch entries.
+            // If we take a snapshot of the epoch entries here and flush to disk outside the read lock,
+            // the leader epoch file may have correctness issue, because the new epoch entry may already in the file.
```

Comment:
Perhaps rewording to sth like the following?

If we take a snapshot of the epoch entries here and flush them to disk outside the read lock, by the time of flushing, the leader epoch file may already be updated with newer epoch entries. Those newer entries will then be overridden with the old snapshot.

Reply:
Thanks for the review and suggestion. Updated it.